### PR TITLE
Cleanup some UxTheme Draw* interop and add tests

### PR DESCRIPTION
**Pull request by hughbe**
## Proposed Changes
- Cleanup UxTheme interop for drawing
- Add VisualStyleRenderer tests

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/2033)

### CODE REVIEW

File: src/Common/src/Interop/UxTheme/Interop.DrawThemeEdge.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    public static partial class UxTheme
+    {
+        [DllImport(Libraries.UxTheme, ExactSpelling = true)]
+        public static extern HRESULT DrawThemeEdge(IntPtr hTheme, IntPtr hdc, int iPartId, int iStateId, ref RECT pDestRect, User32.EDGE uEdge, User32.BF uFlags, ref RECT pContentRect);
+
+        public static HRESULT DrawThemeEdge(HandleRef hTheme, HandleRef hdc, int iPartId, int iStateId, ref RECT pDestRect, User32.EDGE uEdge, User32.BF uFlags, ref RECT pContentRect)
```
Comment (zsd4yr) [inline src/Common/src/Interop/UxTheme/Interop.DrawThemeEdge.cs]: technically can't uEdge be either an EDGE or any of the sunken BDRs?

File: src/Common/src/Interop/User32/Interop.BF.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+internal static partial class Interop
+{
+    internal static partial class User32
```
Comment (zsd4yr) [inline src/Common/src/Interop/User32/Interop.BF.cs]: Are these meant to be in User32 or UxTheme?

Comment (hughbe) [reply src/Common/src/Interop/User32/Interop.BF.cs]: Found in winuser.h

File: src/Common/src/Interop/UxTheme/Interop.DrawThemeEdge.cs (hunk 1 above)
Comment (hughbe) [reply src/Common/src/Interop/UxTheme/Interop.DrawThemeEdge.cs]: EDGE includes the values found in BDR I think

Comment (zsd4yr) [reply src/Common/src/Interop/UxTheme/Interop.DrawThemeEdge.cs]: I am not clear on how that could be the case :( Am I missing something?

Comment (hughbe) [reply src/Common/src/Interop/UxTheme/Interop.DrawThemeEdge.cs]: Ah yes I got that a bit wrong. I've updated `EDGE` to actually include the `BDR_` values listed in https://docs.microsoft.com/en-us/windows/win32/api/uxtheme/nf-uxtheme-drawthemeedge

File: src/Common/src/Interop/User32/Interop.EDGE.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+using System;
+
+internal static partial class Interop
+{
+    internal static partial class User32
+    {
+        [Flags]
+        public enum EDGE : uint
+        {
+            RAISEDOUTER = 0x0001,
```
Comment (zsd4yr) [inline src/Common/src/Interop/User32/Interop.EDGE.cs]: Would is be preferable to do these as

```
RAISEDOUTER = BDR.RAISEDOUTER,
```

**Comment by zsd4yr** [comment]
@msftbot please merge if @RussKie approves

File: src/Common/src/Interop/UxTheme/Interop.IsThemePartDefined.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+using System;
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    public static partial class UxTheme
+    {
+        [DllImport(Libraries.UxTheme, ExactSpelling = true)]
+        public static extern bool IsThemePartDefined(IntPtr hTheme, int iPartId, int iStateId);
```
Comment (RussKie) [inline src/Common/src/Interop/UxTheme/Interop.IsThemePartDefined.cs]:
```suggestion
        public static extern BOOL IsThemePartDefined(IntPtr hTheme, int iPartId, int iStateId);
```

File: src/Common/src/Interop/UxTheme/Interop.IsAppThemed.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+using System;
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    public static partial class UxTheme
+    {
+        [DllImport(Libraries.UxTheme, ExactSpelling = true)]
+        public static extern bool IsAppThemed();
```
Comment (RussKie) [inline src/Common/src/Interop/UxTheme/Interop.IsAppThemed.cs]:
```suggestion
        public static extern BOOL IsAppThemed();
```

File: src/Common/tests/CommonTestHelper.cs [hunk 6]
```diff
@@ -65,6 +65,22 @@ public static TheoryData<Enum> GetEnumTypeTheoryDataInvalid(Type enumType)
             data.Add((Enum)Enum.ToObject(enumType, Convert.ToInt32(values.Max()) + 1));
             return data;
         }
+        public static TheoryData<Enum> GetEnumTypeTheoryDataInvalidMasked(Type enumType)
```
Comment (RussKie) [inline src/Common/tests/CommonTestHelper.cs]: missing new line

File: src/System.Windows.Forms/src/System/Windows/Forms/VisualStyles/EdgeEffects.cs [hunk 7]
```diff
@@ -2,18 +2,17 @@
...
     public enum EdgeEffects
     {
         None = 0,
-        FillInterior = 0x0800,
-        Flat = 0x1000,
-        Soft = 0x4000,
-        Mono = 0x8000,
-        //	#define BF_SOFT         0x1000  /* For softer buttons */
-        //	#define BF_FLAT         0x4000  /* For flat rather than 3D borders */
-        //	#define BF_MONO         0x8000  /* For monochrome borders */
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/VisualStyles/EdgeEffects.cs]: Please migrate these comments into BDR

Comment (hughbe) [reply src/System.Windows.Forms/src/System/Windows/Forms/VisualStyles/EdgeEffects.cs]: They really didn't seem necessary to me

File: src/Common/tests/CommonTestHelper.cs (hunk 6 above)
Comment (hughbe) [reply src/Common/tests/CommonTestHelper.cs]: fixed

File: src/Common/src/Interop/UxTheme/Interop.IsAppThemed.cs (hunk 5 above)
Comment (hughbe) [reply src/Common/src/Interop/UxTheme/Interop.IsAppThemed.cs]: fixed

File: src/Common/src/Interop/UxTheme/Interop.IsThemePartDefined.cs (hunk 4 above)
Comment (hughbe) [reply src/Common/src/Interop/UxTheme/Interop.IsThemePartDefined.cs]: fixed

File: src/System.Windows.Forms/src/System/Windows/Forms/VisualStyles/VisualStyleRenderer.cs [hunk 8]
```diff
@@ -289,14 +288,16 @@ internal void DrawBackground(IDeviceContext dc, Rectangle bounds, IntPtr hWnd)
                 HandleRef hdc = new HandleRef(wgr, wgr.WindowsGraphics.DeviceContext.Hdc);
                 if (IntPtr.Zero != hWnd)
                 {
-                    using (ThemeHandle hTheme = ThemeHandle.Create(_class, true, new HandleRef(null, hWnd)))
+                    using (ThemeHandle hTheme = ThemeHandle.Create(_class, true, hWnd))
                     {
-                        lastHResult = SafeNativeMethods.DrawThemeBackground(new HandleRef(this, hTheme.NativeHandle), hdc, part, state, new NativeMethods.COMRECT(bounds), null);
+                        RECT rect  = bounds;
+                        lastHResult = (int)UxTheme.DrawThemeBackground(new HandleRef(this, hTheme.NativeHandle), hdc, part, state, ref rect, null);
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms/src/System/Windows/Forms/VisualStyles/VisualStyleRenderer.cs]: This shouldn't be creating a HandleRef as the handle isn't stored in "this".

File: src/System.Windows.Forms/src/System/Windows/Forms/VisualStyles/VisualStyleRenderer.cs [hunk 9]
```diff
@@ -289,14 +288,16 @@ internal void DrawBackground(IDeviceContext dc, Rectangle bounds, IntPtr hWnd)
...
                     }
                 }
                 else
                 {
-                    lastHResult = SafeNativeMethods.DrawThemeBackground(new HandleRef(this, Handle), hdc, part, state, new NativeMethods.COMRECT(bounds), null);
+                    RECT rect  = bounds;
+                    lastHResult = (int)UxTheme.DrawThemeBackground(new HandleRef(this, Handle), hdc, part, state, ref rect, null);
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms/src/System/Windows/Forms/VisualStyles/VisualStyleRenderer.cs]: Ideally we should implement `IHandle` so we can avoid creating a `HandleRef`. This particular case is really wonky as `VisualStyleRenderer` doesn't actually have a finalizer. The handle is rooted in the static cache- which should never get collected and therefore the `ThemeHandle` finalizers should never get hit. Yuck.

In any case, we want to keep finalizers from running that would close the handle we're trying to use. As the finalizer is on `ThemeHandle` we should probably be trying to use that directly.

Don't do any of that in this change- I'd just suggest creating a low pri issue to investigate `ThemeHandle` lifetime.

File: src/System.Windows.Forms/src/System/Windows/Forms/VisualStyles/VisualStyleRenderer.cs (hunk 8 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/VisualStyles/VisualStyleRenderer.cs]: As `ThemeHandle` has a finalizer we should actually use it as the owner.

File: src/System.Windows.Forms/src/System/Windows/Forms/VisualStyles/VisualStyleRenderer.cs [hunk 10]
```diff
@@ -329,14 +330,18 @@ internal void DrawBackground(IDeviceContext dc, Rectangle bounds, Rectangle clip
                 HandleRef hdc = new HandleRef(wgr, wgr.WindowsGraphics.DeviceContext.Hdc);
                 if (IntPtr.Zero != hWnd)
                 {
-                    using (ThemeHandle hTheme = ThemeHandle.Create(_class, true, new HandleRef(null, hWnd)))
+                    using (ThemeHandle hTheme = ThemeHandle.Create(_class, true, hWnd))
                     {
-                        lastHResult = SafeNativeMethods.DrawThemeBackground(new HandleRef(this, hTheme.NativeHandle), hdc, part, state, new NativeMethods.COMRECT(bounds), new NativeMethods.COMRECT(clipRectangle));
+                        RECT rect = bounds;
+                        RECT clipRect = clipRectangle;
+                        lastHResult = (int)UxTheme.DrawThemeBackground(new HandleRef(this, hTheme.NativeHandle), hdc, part, state, ref rect, &clipRect);
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms/src/System/Windows/Forms/VisualStyles/VisualStyleRenderer.cs]: The `ThemeHandle` has the finalizer so it should be kept alive. 'this' won't do it.

File: src/System.Windows.Forms/src/System/Windows/Forms/VisualStyles/VisualStyleRenderer.cs [hunk 11]
```diff
@@ -468,7 +471,8 @@ public void DrawParentBackground(IDeviceContext dc, Rectangle bounds, Control ch
                 using (WindowsGraphicsWrapper wgr = new WindowsGraphicsWrapper(dc, AllGraphicsProperties))
                 {
                     HandleRef hdc = new HandleRef(wgr, wgr.WindowsGraphics.DeviceContext.Hdc);
-                    lastHResult = SafeNativeMethods.DrawThemeParentBackground(new HandleRef(this, childControl.Handle), hdc, new NativeMethods.COMRECT(bounds));
+                    RECT rc = bounds;
+                    lastHResult = (int)UxTheme.DrawThemeParentBackground(new HandleRef(this, childControl.Handle), hdc, ref rc);
```
Comment (JeremyKuhne) [inline src/System.Windows.Forms/src/System/Windows/Forms/VisualStyles/VisualStyleRenderer.cs]: `childControl` is the class that has the finalizer for `childControl.Handle`. It is what should be kept alive, not `this`.

File: src/System.Windows.Forms/src/System/Windows/Forms/VisualStyles/EdgeEffects.cs [hunk 12]
```diff
@@ -2,18 +2,17 @@
...
+        FillInterior = (int)User32.BF.MIDDLE,
+        Flat = (int)User32.BF.FLAT,
+        Soft = (int)User32.BF.SOFT,
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/VisualStyles/EdgeEffects.cs]: ❗️ We have a breaking change here.

| Const | Original | New |
|---|---|---|
| `Flat` | 0x1000 | `BF.FLAT` = 0x4000 |
| `Soft` | 0x4000 | `BF.SOFT` = 0x1000 |

Granted the C++ definitions are as were noted in the comments and align with the new values:
image

However the original bug was likely introduced in .NET 2.0 timeframe, and thus has been around for a very long time.

/cc: @merriemcgaw @Tanya-Solyanik

File: src/System.Windows.Forms/src/System/Windows/Forms/Border3DStyle.cs [hunk 13]
```diff
@@ -16,51 +17,51 @@ public enum Border3DStyle
...
         ///  The border has a with a sunken inner edge and no outer edge.
         /// </summary>
-        SunkenInner = NativeMethods.BDR_SUNKENINNER,
+        SunkenInner = (int)User32.BDR.SUNKENINNER,
 
         /// <summary>
         ///  The border has a with a sunken outer edge and no inner edge.
         /// </summary>
-        SunkenOuter = NativeMethods.BDR_SUNKENOUTER,
+        SunkenOuter = (int)User32.BDR.SUNKENOUTER,
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Border3DStyle.cs]: I don't believe the definition is correct, although the underlying values are the same.
The name of the enum is "Border3D", so it must reference the border enum, which is `BDR` not `EDGE` or `BF`.

**Pull request merged**